### DISCUSSION TITLE
Unexpected behaviour with Matplotlib 3.8.0.

### DISCUSSION THREAD

**Discussion by KelSolaar** [microsoft/pyright#5967] [Q&A] [answered]
Hello,

Matplotlib 3.8.0 is generating a ton of errors on CI where 3.7.2 was fine.

```python
import matplotlib.pyplot as plt


def foo()-> tuple[plt.Axes, plt.Figure]:
    ...
```

```bash
$ pyright --version
pyright 1.1.327
$ python -c "import matplotlib;print(matplotlib.__version__)"
3.8.0
$ pyright "scratch_104.py"
scratch_104.py
  scratch_104.py:4:23 - error: "Axes" is not a known member of module "matplotlib" (reportGeneralTypeIssues)
  scratch_104.py:4:33 - error: "Figure" is not a known member of module "matplotlib" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
```

I'm unclear as to why `Axes` and `Figure` are not picked up as they are imported in `matplotlib/pyplot.py`:
https://github.com/matplotlib/matplotlib/blob/58d5aad7ba8bec31cf1ff5dc4a758fabc9b7fe16/lib/matplotlib/pyplot.py#L66

Cheers,

Thomas

**Comment by erictraut** [comment, marked as answer]
It looks like matplotlib 3.8.0 includes a "py.typed" marker file, whereas previous versions do not. When this marker file is present, it means that type checkers and other static analysis tools should assume that the package is following certain rules as documented [here](https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface-public-and-private-symbols). By these rules, `Axes` and `Figure` are considered private imports, not intended to be re-exported by the `matplotlib.pyplot` submodule. If the library maintainers intend for these symbols to be re-exported, they would need to use a redundant form of the import statement, as in `from matplotlib.axes import Axes as Axes`. This convention was initially specified in PEP 484 for type stubs, but it was later adopted for "py.typed" libraries that provide inlined type information.

If you think this is a bug in the library, please file an issue with the matplotlib maintainers.

**Reply by KelSolaar** [reply]
Ah fantastic @erictraut, I did not know that document you linked this is great resource.

As a quick test, I added the `Axes` and `Figure` object to a new `__all__` attribute in the `pyplot` module and it got rid of the errors:

`__all__ = ["Axes", "Figure"]`


I will get in touch with the MPL developers!

**Reply by tacaswell** [reply]
Those classes are not meant to be re-exported from `matplotlib.pyplot`.  I think this is an example of typing correctly catching unexpected usage by down-stream libraries.
